### PR TITLE
Add observability for inline wrapping token classification (#296)

### LINKED ISSUE #296: Add observability infrastructure for inline-wrapping token classification
## Description

`AGENTS.md` requires structured diagnostics (`tracing`) and operational metrics (`metrics`) for behaviour-affecting code paths. The GFM footnote-reference work introduced in PR #280 adds new token-classification and fragment-classification paths but includes no `tracing::debug` spans, instrumentation, or metrics. This issue tracks the work needed to bring the inline-wrapping pipeline into compliance with the project observability policy.

## Affected code paths (non-exhaustive)

- `parse_link_or_image` / `find_footnote_end` in `src/wrap/tokenize/parsing.rs` — footnote-reference detection and early return.
- `classify_fragment` / `looks_like_footnote_ref` / `ends_with_footnote_ref` in `src/wrap/inline/fragment.rs` — fragment-kind classification.
- `determine_token_span` / `promote_footnote_span` in `src/wrap/inline.rs` — span-kind promotion during grouping.
- Any further token-classification or fragment-classification paths added in future PRs that alter wrap behaviour.

## Required work

1. **Tracing instrumentation** — Add `tracing::debug!` (or `tracing::trace!`) calls at decision points where a footnote reference (or other atomic token) is recognised, emitted, or promoted. Use field-based filtering (e.g. `token = %token`, `kind = ?kind`) rather than string interpolation, in accordance with `AGENTS.md`.
2. **Metrics** — Evaluate whether a counter for footnote tokens recognised per wrap call (and per document) is meaningful for operational use. If so, emit it via the `metrics` façade with a stable name and low-cardinality labels.
3. **Scope** — Apply the same pattern to all present and future token-classification paths in the wrapping pipeline, not only the footnote path. Establish a consistent instrumentation convention so new paths are covered by default.
4. **No global recorder/subscriber installation** — Libraries must not install global `tracing` subscribers or `metrics` recorders; instrumentation must be purely additive.

## Acceptance criteria

- All classification decision points in `src/wrap/tokenize/` and `src/wrap/inline/` emit at least one `tracing` event at `DEBUG` or `TRACE` level when a non-trivial classification is made.
- `make lint` and `make test` pass with no warnings after the changes.
- No global subscriber or recorder is installed by the library.

## References

- PR #280 (GFM footnote-reference atomicity)
- `AGENTS.md` — Observability section

---

Reported in PR #280 by @leynos.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

Closes #296

This branch brings the inline-wrapping token- and fragment-classification
pipeline into compliance with the project observability policy introduced
for GFM footnote-reference handling in #280.

- Adds the `tracing` crate and derives `Debug` on `SpanKind` so
  classification outcomes appear as structured fields.
- Emits `debug!` at high-value decision points (footnote parse, fragment
  classification, span promotion) and `trace!` at branch-level checks,
  using consistent field names (`token`, `kind`, `start`, `end`, `width`).
- Defers operational `metrics` counters until a clear per-wrap use case
  emerges; instrumentation remains purely additive with no global
  subscriber or recorder installation.

## Test plan

- [x] `make check-fmt`
- [x] `make lint`
- [x] `make test`
- [x] CodeRabbit review (`coderabbit review --agent`) — 0 findings

Made with [Cursor](https://cursor.com)

## Summary by Sourcery

Add structured tracing around inline wrapping tokenization and classification to improve observability without changing wrapping behavior.

Enhancements:
- Introduce structured debug/trace logging for footnote detection, fragment classification, token span grouping, and link/image parsing in the inline wrap pipeline.
- Derive Debug for span classification types and add helpers to standardize and truncate logged fields for trace output.

Build:
- Add the tracing crate as a dependency for observability instrumentation.

### CODE REVIEW

File: src/wrap/inline/fragment.rs [hunk 1]
```diff
@@ -112,25 +113,42 @@ pub(super) fn width_as_f64(width: usize) -> f64 {
...
-    {
-        FragmentKind::FootnoteRef
+/// Returns `text` truncated for structured tracing when it exceeds 80 columns.
+fn fragment_text_for_trace(text: &str) -> String {
+    const MAX_TRACE_TEXT: usize = 80;
+    if text.len() <= MAX_TRACE_TEXT {
+        text.to_string()
     } else {
-        FragmentKind::Plain
+        format!("{}…", &text[..MAX_TRACE_TEXT])
```
[the remark this answers is not in the document]
Comment (leynos) [reply src/wrap/inline/fragment.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 errors, 7 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | No tests verify the new tracing instrumentation. Existing tests assert only classification results, not tracing calls; they would pass identically if all trace!() and debug!() calls were removed. | Add tests that capture and verify trace/debug output at decision points using tracing\_test or similar; verify calls occur for footnote classification, link parsing, span promotion. |
| Unit Architecture | ❌ Error | Query functions emit tracing side-effects but do not expose this in signatures, names, or return types, hiding side-effects behind apparently pure read APIs. | Make side-effects explicit: inject tracing dependency as parameter, use #\[tracing::instrument\] at public boundaries, or defer instrumentation to dedicated wrapper functions. |
| Developer Documentation | ⚠️ Warning | PR adds new `tracing` build requirement and instrumentation patterns to wrapping pipeline, but `docs/developers-guide.md` lacks documentation of the observability conventions per custom check. | Document the new `tracing` dependency, instrumentation pattern, emitted fields (token, kind, start, end, width), and constraint that library does not install global subscribers in developers-guide.md. |
| Testing (Unit And Behavioural) | ⚠️ Warning | No tests added for the instrumentation. Review comments flag UTF-8 slicing panic in fragment\_text\_for\_trace and code complexity issues that require testing and refactoring. | Add unit tests for UTF-8 edge cases in fragment\_text\_for\_trace. Extract classification predicates to reduce complexity. Verify file length compliance and integration test coverage. |
| Testing (Property / Proof) | ⚠️ Warning | PR introduces unsafe UTF-8 truncation with unchecked byte indexing; proptest property tests are needed to verify invariant that panics never occur on arbitrary input. | Add proptest verifying `fragment_text_for_trace()` never panics on arbitrary UTF-8; or fix unsafe slicing using boundary-safe `chars().take(80)` truncation. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | PR adds tracing instrumentation but omits snapshot tests. The workspace has insta for verifying structured output; tracing events should be tested. | Add snapshot tests using insta to verify tracing events emit correctly with field values (token, kind, start, end) for token, fragment, and footnote classification paths. |
| Observability | ⚠️ Warning | Fragment\_text\_for\_trace uses unsafe UTF-8 byte-slicing that panics on multi-byte characters, compromising logging reliability in production. | Use char-safe truncation instead of byte-indexing; extract inline.rs logging helpers; refactor classify\_fragment to predicate helpers. |
| Performance And Resource Use | ⚠️ Warning | Unconditional String allocations in hot path (fragment\_text\_for\_trace per fragment); UTF-8 panic risk with &text\[..80\] on multibyte boundaries; missing span\_enabled() guards. | Guard fragment\_text\_for\_trace() call with span\_enabled!(); fix UTF-8 slicing via character-safe truncation; apply conditional guards to all expensive field expressions. |
| Architectural Complexity And Maintainability | ⚠️ Warning | src/wrap/inline.rs exceeds 400-line limit (493 lines). Logging helpers must extract to submodule. UTF-8 slicing in fragment\_text\_for\_trace risks panic. | Extract logging helpers to dedicated submodule. Replace &text\[..MAX\_TRACE\_TEXT\] with chars().take() for UTF-8-safe truncation. |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Property / Proof) | ⚠️ Warning | Property test trace\_text\_snippet\_never\_panics only verifies char-boundary safety; omits assertions for documented invariants "≤80 bytes" and "truncation flag", making it incomplete. | Add length and truncation-flag assertions to property test to cover all documented invariants: snippet.len() ≤ 80 and verify truncation boolean accuracy. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | PR lacks compile-time tests (trybuild) for #\[tracing::instrument\] attributes and snapshot tests for structured observability output. | Add trybuild tests verifying #\[tracing::instrument\] attributes compile, and add snapshot tests (insta) capturing structured tracing output. |

File: tests/ui/tracing_instrument_pass.rs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+//! Compile-pass fixture: #[tracing::instrument] is accepted on free functions
+//! that return bool and on functions that return Option<usize>.
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/ui/tracing_instrument_pass.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (3 errors, 4 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | The docstring in tests/ui/tracing\_instrument\_pass.rs claims to test functions returning both bool and Option, but only tests bool returns. | Update the fixture docstring to remove the mention of Option, or add test functions returning Option. |
| Module-Level Documentation | ❌ Error | The module docstring in tests/ui/tracing\_instrument\_pass.rs incorrectly claims coverage of functions returning Option, but the module contains only functions returning bool. | Update the docstring in tests/ui/tracing\_instrument\_pass.rs to: "//! Compile-pass fixture: #\[tracing::instrument\] is accepted on free functions that return bool." |
| Unit Architecture | ❌ Error | The docstring in tests/ui/tracing\_instrument\_pass.rs claims coverage of Option returns, but only bool functions exist in the fixture. | Remove the reference to "functions that return Option" from the fixture's module docstring at line 2. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Fragment tracing tests lack assertions around logs\_contain() calls; predicates with #\[tracing::instrument\] have no traced\_test coverage; UI fixture docstring incorrectly mentions Option. | Fix fragment test assertions to use assert!(logs\_contain()); add traced\_test cases for looks\_like\_footnote\_ref and ends\_with\_footnote\_ref; correct fixture docstring to mention only bool. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | The fixture docstring claims to test functions returning "bool and Option" but only tests bool return types, violating the requirement to encode meaningful behavioural expectations accurately. | Update tests/ui/tracing\_instrument\_pass.rs line 1–2 docstring to "accept on free functions that return bool", matching the actual implementations tested. |
| Observability | ⚠️ Warning | The docstring in tests/ui/tracing\_instrument\_pass.rs is inaccurate. It claims functions return "bool and on functions that return Option" but both fixture functions return only bool. | Update lines 1-2 of tests/ui/tracing\_instrument\_pass.rs to remove "and on functions that return Option" from the module docstring to match actual implementations. |
| Architectural Complexity And Maintainability | ⚠️ Warning | The fixture docstring in tests/ui/tracing\_instrument\_pass.rs incorrectly claims coverage for Option when only bool return types are tested, creating documentation debt. | Update the docstring to match actual coverage: remove the "and on functions that return Option" clause. |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests use only string-matching and would not catch incorrect field values, inverted conditions, or logic bypass. Removing tracing::enabled!() guards would not cause test failures. | Strengthen tests to assert structured field values via logs\_contain patterns or snapshot testing; add tests for tracing::enabled!() guards and log-level changes. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Tests verify logging occurs but lack meaningful coverage: only 3 of 5 fragment types tested; assertions check substrings not structured field values. | Cover all 5 fragment types; validate structured field values (token, kind, truncated); test truncation edge cases. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Trybuild compile-pass test present for #\[tracing::instrument\] attributes, but snapshot tests for structured tracing output are absent. Current tests only verify event emission, not output structure. | Add insta snapshot tests capturing structured tracing event output (token, kind, truncated fields) to encode meaningful behavioural expectations and prevent incidental output churn. |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests use only string-matching and would not catch incorrect field values, inverted conditions, or logic bypass. Removing tracing::enabled!() guards would not cause test failures. | Strengthen tests to assert structured field values via logs\_contain patterns or snapshot testing; add tests for tracing::enabled!() guards and log-level changes. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Tests verify logging occurs but lack meaningful coverage: only 3 of 5 fragment types tested; assertions check substrings not structured field values. | Cover all 5 fragment types; validate structured field values (token, kind, truncated); test truncation edge cases. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Trybuild compile-pass test present for #\[tracing::instrument\] attributes, but snapshot tests for structured tracing output are absent. Current tests only verify event emission, not output structure. | Add insta snapshot tests capturing structured tracing event output (token, kind, truncated fields) to encode meaningful behavioural expectations and prevent incidental output churn. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- The new tracing in `classify_fragment` calls `fragment_text_for_trace`, which allocates a `String` for every fragment even when logging is disabled; consider logging `&str` directly and letting the subscriber handle truncation, or returning a `Cow<'_ ,str>` to avoid allocations on the hot path.
- Several trace/debug statements (e.g., in `find_footnote_end` and `parse_link_or_image`) construct owned `String`s via `collect_range` purely for logging; you may want to log slices (`&str`) or defer allocation behind `debug_enabled!()`/`trace_enabled!()` checks to reduce overhead in tight parsing loops.

## Individual Comments

### Comment 1
<location path="src/wrap/inline/fragment.rs" line_range="122" />
<code_context>
+        text.to_string()
     } else {
-        FragmentKind::Plain
+        format!("{}…", &text[..MAX_TRACE_TEXT])
     }
 }
</code_context>
<issue_to_address>
**issue (bug_risk):** Using a fixed byte index to slice `&str` can panic on non-ASCII input.

`text[..MAX_TRACE_TEXT]` slices by byte index, so if `text` has multi-byte chars and the cutoff falls inside one, this will panic. Please truncate on a character or grapheme boundary instead (e.g., via `char_indices` or `unicode-segmentation`) to ensure valid UTF-8 slices.
</issue_to_address>

### Comment 2
<location path="src/wrap/inline/fragment.rs" line_range="115" />
<code_context>
 /// still recognised as links or code spans. Footnote references also recognise
 /// the `word.[^label]` suffix shape that the wrapper groups to avoid splitting
 /// sentence punctuation from the marker.
-fn classify_fragment(text: &str) -> FragmentKind {
-    if is_whitespace_token(text) {
-        return FragmentKind::Whitespace;
</code_context>
<issue_to_address>
**issue (review_instructions):** `classify_fragment` both performs classification and logging, and still embeds a multi-branch conditional instead of delegating the branchy logic into a predicate helper.

This function is now doing two things: computing the `FragmentKind` and emitting structured tracing. To better follow single-responsibility/CQRS, consider keeping `classify_fragment` focused on determining the kind (pure classification) and moving the logging into a thin wrapper or caller, or into a dedicated helper that takes the already-computed kind.

Additionally, the nested `if / else if` chain that selects the `FragmentKind` has more than two branches. Per the guideline about conditionals with >2 branches, it would be better to extract this branching logic into one or more predicate helpers (e.g. `is_inline_code_fragment(..)`, `is_link_fragment(..)`, etc.) rather than keeping all the cases inline here.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
* Adhere to single responsibility and CQRS
* Move conditionals with >2 branches into a predicate function.

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### Assessment against linked issues

| Issue | Objective | Addressed | Explanation |
| --- | --- | --- | --- |
| [#296](https://github.com/leynos/mdtablefix/issues/296) | Add structured tracing (debug/trace) events at all non-trivial classification decision points in the inline wrapping pipeline (src/wrap/tokenize/, src/wrap/inline/, src/wrap/inline/fragment.rs), including footnote and other token/fragment classifications, using field-based logging and without installing a global tracing subscriber. | ✅ |  |
| [#296](https://github.com/leynos/mdtablefix/issues/296) | Evaluate the use of operational metrics (e.g., counters for recognised footnote tokens per wrap/document) and ensure any observability changes remain purely additive, with no global metrics recorder or tracing subscriber installed. | ✅ |  |
| [#296](https://github.com/leynos/mdtablefix/issues/296) | Establish a consistent observability convention for token and fragment classification (including non-footnote paths) so that present and future classification code paths can be instrumented uniformly. | ✅ |  |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (3 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Unit And Behavioural) | ⚠️ Warning | Tests directly exercise private trace\_text\_snippet helper; tracing tests verify instrumentation rather than functional correctness; no e2e tests confirm wrapping still works correctly. | Remove tests of private trace\_text\_snippet function; refactor tracing tests to verify public function correctness, not log emissions; add e2e tests ensuring wrapping pipeline produces correct output with instrumentation. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Trybuild test present, but snapshot tests for structured tracing output deferred to `#305` instead of included per check requirements for text-based/structured output. | Include insta snapshot tests capturing structured tracing events' complete serialisation, or expand this PR scope to include `#305`'s snapshot testing work. |
| Observability | ⚠️ Warning | Code uses undocumented fields (`reason`, `is_image`) that violate the stated "stable structured field names" requirement in developers-guide.md. | Update developers-guide.md to document additional context-specific fields, or refactor code to use only documented field names (token, kind, start, end, width, truncated). |

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

### ❌ Failed checks (1 error, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | find\_footnote\_end emits 3 trace events but tests cover only 1 path; footnote\_label\_span\_recognised and unterminated\_bracket branches are not tested. | Add tests for all find\_footnote\_end branches: happy path (footnote\_label\_span\_recognised) and unterminated bracket case to verify tracing emission. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Unit tests verify local behaviour but lack end-to-end tests exercising observable tracing behaviour through the library's public wrapping API. | Add integration tests that call wrap\_text with footnoted content and use #\[traced\_test\] to verify classification debug events are emitted at the functional boundary. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 4 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Unit Architecture | ❌ Error | Query predicates emit tracing side-effects via global context without visible signatures; tracing dependencies are hidden rather than injected. | Wrap instrumented predicates in command functions, inject tracing context as a parameter, or move instrumentation to callers to make side-effects explicit and optional. |
| Developer Documentation | ⚠️ Warning | Unaddressed review comment: docs say "sole observability dependency" not "sole runtime observability dependency" and omit tracing-test, misaligning with Cargo.toml. | Change line 296 to "sole runtime observability dependency" and add sentence about tracing-test in dev-dependencies to match Cargo.toml. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Unit tests verify event emission with edge cases but lack integration tests. No end-to-end tests exercise the full wrapping pipeline with tracing whilst verifying outputs remain correct. | Add integration tests using traced\_test for the classification pipeline to verify instrumentation preserves functional correctness. |
| Observability | ⚠️ Warning | Observability is comprehensive with structured tracing, allocation guards and tests. The developers-guide Observability section fails to clarify that tracing-test is a test-only dev-dependency. | Add to developers-guide.md line 297: "Test-only instrumentation uses tracing-test under dev-dependencies." to clarify test-only dependency scope per review comment. |
| Performance And Resource Use | ⚠️ Warning | Hot-path debug/trace calls in parsing.rs lack tracing::enabled! guards; field formatters evaluate unconditionally despite fragment.rs correctly guarding equivalent expressions. | Wrap debug!/trace! field value expressions in parse\_link\_or\_image and find\_footnote\_end with if tracing::enabled!() guards, matching the pattern in log\_fragment\_classification. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (4 errors, 1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Tests for tracing are vacuous: they only verify log names exist via substring search, not that structured fields and values are correct or logged under right conditions. | Replace logs\_contain() substring checks with logs\_assert() to verify complete structured fields, values, and emission conditions; add negative tests. |
| Module-Level Documentation | ❌ Error | Test submodules in fragment.rs, parsing.rs, and predicates.rs lack module-level //! doc comments. AGENTS.md requires every module to begin with //! explaining its purpose. | Add //! doc comments to trace\_snippet\_tests, tracing\_tests, and other test submodules explaining their purpose and role in validating instrumentation and trace event emission. |
| Unit Architecture | ❌ Error | #\[tracing::instrument\] hides side-effects in query functions; violates requirement that side-effects be explicit in names, types, and signatures. | Replace #\[tracing::instrument\] on queries with explicit trace calls guarded by tracing::enabled!, making side-effects visible in function bodies. |
| Security And Privacy | ❌ Error | URL tokens logged at DEBUG level without truncation risk exposing embedded secrets (API keys, tokens, IDs). Documentation lacks redaction guidance for sensitive fields. | Add guidance to docs/developers-guide.md recommending subscribers filter/redact the `token` field when enabling DEBUG logging in production environments. |
| Testing (Compile-Time / Ui) | ⚠️ Warning | Trybuild test for compile-time behaviour present and correct; however, snapshot tests for structured trace output are absent and should be added per custom check guidance. | Add snapshot tests encoding the exact structure of trace events (field names, values) using insta, or document in PR why snapshot testing is deferred and under which issue number. |

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

Domain Architecture ⚠️ Warning Domain functions directly import and use vendor-specific tracing framework (tracing::instrument, debug!, enabled!). These infrastructure concerns should be isolated at adapter boundaries. Define domain-level event types (e.g. enum ClassificationEvent) and pass optional observer into classification functions; emit domain events at boundaries, not vendor-specific tracing inside domain logic.
